photon-api/src/main/scala/com/linkedin/photon/ml/data/LocalDataSet.scala

```diff
@@ -176,6 +218,105 @@ object LocalDataSet {
     result
   }
 
+  /**
+   * Compute feature ratio confidence interval lower bounds.
+   *
+   * @param labelAndFeatures An [[Array]] of (label, feature vector) tuples
+   * @param zScore The Z-score for the chosen two-tailed confidence level
+   * @param globalFeatureInstances The global instances with the features present
+   * @param globalPositiveInstances The global positive instances with the features present
+   * @param binaryIndices The binary feature columns indices
+   * @param epsilon The constant used to compute for extreme case of ratio modeling
+   * @return the lowerBounds for feature columns
+   */
+  protected[ml] def computeRatioCILowerBound(
+      labelAndFeatures: Array[(Double, Vector[Double])],
+      globalFeatureInstances: Array[Double],
+      globalPositiveInstances: Array[Double],
+      binaryIndices: Set[Int],
+      numFeatures: Int,
+      zScore: Double): Map[Int, Double] = {
+
+    val n = globalFeatureInstances
+    val y = globalPositiveInstances
+
+    val m = labelAndFeatures.map(_._2).reduce(_ + _)
+    val x = labelAndFeatures
+      .filter(_._1 > MathConst.POSITIVE_RESPONSE_THRESHOLD)
+      .map(_._2)
+      .foldLeft(Vector.zeros[Double](numFeatures))(_ + _)
+
+    binaryIndices
+      .map { key =>
+        val x_col = x(key)
+        val m_col = m(key)
+        val y_col = y(key)
+        val n_col = n(key)
+
+        val lowerBound = if (y_col == 0.0 || y_col == n_col) {
+          0D
```

Comment:
for m_col = 0, we may also want to return 0D here. Since if it enter next branch it will produce Infinity.